group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -892,6 +893,43 @@ public void replay(
         }
     }
 
+    /**
+     * Applies the given transaction marker.
+     *
+     * @param producerId    The producer id.
+     * @param result        The result of the transaction.
+     * @throws RuntimeException if the transaction can not be completed.
+     */
+    public void completeTransaction(
```

Comment:
When the marker is written, the state machine moves the pending offsets of the producer id to the main store if the transaction was committed; there are just removed otherwise.